### PR TITLE
EPD-845: Add Remote Configs

### PR DESCRIPTION
**Pull request by adamnolte**
will add e2e tests once we have it all deployed

### CODE REVIEW

File: src/configs/types.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    id: z.string(),
+    dangerouslyUseUndeployed: z.literal(true),
+  }),
+  z.object({
+    id: z.string(),
+    version: z.string(),
+  }),
+  z.object({
+    id: z.string(),
+    revisionId: z.string(),
```
Comment (nicolewhite) [inline src/configs/types.ts:33]: i think we want pinning to a specific revisionId under the dangerously-use-undeployed umbrella, since you'd only want to use a specific revision during local development and the dangerously-use-* prefix is intended to be a clue to you (and to code reviewers if you accidentally open a pull request w/o backing out your temporary changes) that it's not intended to be used in production. as-is you would be able to do:

```ts
await config.activateRemoteConfig({
  config: {
    id: 'my-config-id',
    revisionId: 'clvifur81000336hksycpw4av',
  },
});
```

which doesn't have any hints that you're about to commit something dangerous (an undeployed revision) to production

so for prompts for example, my plan is to update the manager to be like:

```ts
const mgr = new AutoblocksPromptManager({
  id: 'my-prompt',
  version: {
    // Use latest undeployed revision; requires a user-scoped API key
    major: 'dangerously-use-undeployed',
    minor: 'latest',
  },
});

const mgr = new AutoblocksPromptManager({
  id: 'my-prompt',
  version: {
    // Use specific undeployed revision; requires a user-scoped API key
    major: 'dangerously-use-undeployed',
    minor: 'clvifur81000336hksycpw4av',
  },
});

const mgr = new AutoblocksPromptManager({
  id: 'my-prompt',
  version: {
    // Use deployed revision 1.0
    major: '1',
    minor: '0',
  },
});

const mgr = new AutoblocksPromptManager({
  id: 'my-prompt',
  version: {
    // Use latest deployed revision w/ major version 1
    major: '1',
    minor: 'latest',
  },
});
```

in other words, there is no way to specify a specific revisionId w/o having `dangerously-use-undeployed` there to hint that you're using something dangerous intended only for local dev scenarios

Comment (nicolewhite) [reply src/configs/types.ts:33]: i wonder if it should be like:

```ts
export const zConfigParameterSchema = z.union([
  ...
  z.object({
    id: z.string(),
    // Using this requires a user-scoped API key
    dangerouslyUseUndeployed: z.union([
      z.object({
        // Use latest undeployed revision
        latest: z.literal(true),
      }),
      z.object({
        // Use specific undeployed revision
        revisionId: z.string(),
      }),
    ]),
  }),
  ...
]);
```

?

File: src/configs/types.ts [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+export const zConfigSchema = z.object({
+  id: z.string(),
+  version: z.string(),
+  value: z.record(z.string(), z.unknown()),
+});
+
+export type Config = z.infer<typeof zConfigSchema>;
+
+export enum ConfigSpecialVersion {
```
Comment (nicolewhite) [inline src/configs/types.ts:11]: can you use the one already in `src/util.ts` (and rename it to `SpecialVersion` or something)

https://github.com/autoblocksai/javascript-sdk/blob/060e0a0feaf7b03deebd44653ff8930a55f6ae7e/src/util.ts#L28-L31

File: src/configs/index.ts [hunk 3]
```diff
@@ -0,0 +1 @@
+export { AutoblocksConfig } from './config';
```
Comment (nicolewhite) [inline src/configs/index.ts:1]: do we want this to be AutoblocksConfigManager to match prompts or nah ?

Comment (adamnolte) [reply src/configs/index.ts:1]: I was debating, but decided not to since it works differently

File: src/configs/config.ts [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+
+  constructor(value: T) {
+    this._value = value;
+  }
+
+  private makeRequestUrl(args: { id: string; version: ConfigVersion }): string {
+    const configId = encodeURIComponent(args.id);
+    let version: string;
+    if (args.version === ConfigSpecialVersion.DANGEROUSLY_USE_UNDEPLOYED) {
+      version = 'undeployed';
```
Comment (nicolewhite) [inline src/configs/config.ts:60]: can we pull this out into a constant, ideally pull this one out too and use in both places: https://github.com/autoblocksai/javascript-sdk/blob/060e0a0feaf7b03deebd44653ff8930a55f6ae7e/src/prompts/manager.ts#L56

File: src/configs/types.ts (hunk 1 above)
Comment (adamnolte) [reply src/configs/types.ts:33]: Ya I think that makes sense. I can update to that format

File: src/configs/config.ts [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+
+    return `${API_ENDPOINT}/configs/${configId}/${version}`;
+  }
+
+  private async getConfig(args: {
+    id: string;
+    version: ConfigVersion;
+    timeoutMs: number;
+    apiKey: string;
+  }): Promise<Config | undefined> {
```
Comment (nicolewhite) [inline src/configs/config.ts:73]: isn't this `Promise<Config>`?

Comment (adamnolte) [reply src/configs/config.ts:73]: Oh ya I changed this and forgot to update the return type

File: src/configs/config.ts [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+    const remoteConfig = await this.getConfig({
+      id: args.id,
+      version: args.version,
+      timeoutMs: convertTimeDeltaToMilliSeconds(
+        args.timeout || { seconds: 10 },
+      ),
+      apiKey: args.apiKey,
+    });
+
+    if (!remoteConfig) {
```
Comment (nicolewhite) [inline src/configs/config.ts:104]: i think can drop assuming the return type of `getConfig` is `Config` and not `Config | undefined`

File: src/configs/config.ts [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+    let version: string;
+    if (
+      args.version === RevisionSpecialVersionsEnum.DANGEROUSLY_USE_UNDEPLOYED
+    ) {
+      version = REVISION_UNDEPLOYED_VERSION;
+    } else {
+      version = encodeURIComponent(args.version);
+    }
+
+    return `${API_ENDPOINT}/configs/${configId}/${version}`;
```
Comment (nicolewhite) [inline src/configs/config.ts:63]:
```suggestion
    return `${API_ENDPOINT}/configs/${configId}/versions/${version}`;
```

File: src/configs/config.ts [hunk 8]
```diff
@@ -49,27 +48,40 @@ export class AutoblocksConfig<T> {
...
+      ('dangerouslyUseUndeployed' in args.remoteConfig &&
+        'latest' in args.remoteConfig.dangerouslyUseUndeployed)
+    );
+  }
+
+  private makeRequestUrl(args: { remoteConfig: RemoteConfig }): string {
+    const configId = encodeURIComponent(args.remoteConfig.id);
+    const base = `${API_ENDPOINT}/configs/${configId}`;
+    // Overide the version using the revisionId in the map if it exists
+    const revisionId = configRevisionsMap()[args.remoteConfig.id];
```
Comment (nicolewhite) [inline src/configs/config.ts:63]: in the prompt SDK i'm being super defensive about the overrides and double checking in multiple places that we're in a testing context:

* https://github.com/autoblocksai/javascript-sdk/blob/a499c2114d099e1d77c351088af1bd88a3729057/src/prompts/manager.ts#L316
* https://github.com/autoblocksai/javascript-sdk/blob/a499c2114d099e1d77c351088af1bd88a3729057/src/prompts/manager.ts#L372

should we do the same here? so just:

```ts
if (isTestingContext() && configRevisionsMap()[args.remoteConfig.id]) {
  const revisionId = configRevisionsMap()[args.remoteConfig.id];
  return `${base}/revisions/${revisionId}`;
}
```

just b/c it would be easy to accidentally refactor something like `configRevisionsMap()` to return the revision ID even when not in a testing context or something

but maybe it's overly defensive, lmk

File: test/configs/config.spec.ts [hunk 9]
```diff
@@ -159,7 +159,7 @@ describe('AutoblocksConfig', () => {
       });
 
       await config.activateRemoteConfig({
-        config: {
+        remoteConfig: {
```
Comment (nicolewhite) [inline test/configs/config.spec.ts:162]: hmm i think i liked just config, seems kinda redundant now since the method name is already "activateRemoteConfig"

File: test/configs/config.spec.ts [hunk 10]
```diff
@@ -159,7 +159,7 @@ describe('AutoblocksConfig', () => {
...
           id: 'my-config-id',
```
Comment (nicolewhite) [inline test/configs/config.spec.ts:163]: just occurred to me, shouldn't the config ID be specified when initializing the `AutoblocksConfig` class? it wouldn't make sense to activate a remote config w/ different IDs from the same instance would it?

Comment (nicolewhite) [reply test/configs/config.spec.ts:163]: so like this:

```ts
const config = new AutoblocksConfig<{ foo: string }>({
  id: 'my-config-id',
  values: { foo: 'bar' },
});

await config.activateRemoteConfig({
  version: {
    latest: true,
  },
  apiKey: '...',
});

// or 

await config.activateRemoteConfig({
  version: {
    dangerouslyUseUndeployed: {
      latest: true,
    },
  },
  apiKey: '...',
});
```

Comment (nicolewhite) [reply test/configs/config.spec.ts:163]: oh i see, ya that is true

**Review by nicolewhite — APPROVED** [review]

File: test/configs/config.spec.ts (hunk 9 above)
Comment (adamnolte) [reply test/configs/config.spec.ts:162]: ya i was going back forth on this lol

File: src/configs/config.ts (hunk 8 above)
Comment (adamnolte) [reply src/configs/config.ts]: ya I think that makes sense, I update this check for testing context

File: e2e/test-configs-ts/package.json [hunk 11]
```diff
@@ -0,0 +1,26 @@
+{
+  "name": "test-prompt-manager-ts",
```
Comment (nicolewhite) [inline e2e/test-configs-ts/package.json:2]:
```suggestion
  "name": "test-configs-ts",
```

File: e2e/test-configs-ts/package.json [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+  "version": "0.0.0",
+  "private": true,
+  "engines": {
+    "node": ">=18"
+  },
+  "type": "module",
+  "scripts": {
+    "test": "jest",
+    "gen": "prompts generate",
+    "gen:watch": "nodemon --quiet --watch prompt-templates --ext \"*\" --exec \"npm run gen\""
```
Comment (nicolewhite) [inline e2e/test-configs-ts/package.json:12]: do you need these?

**Review by nicolewhite — APPROVED** [review]

**Review by nicolewhite — APPROVED** [review]
